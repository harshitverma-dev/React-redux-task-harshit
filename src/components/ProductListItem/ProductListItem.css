.product-listing-item {
    display: flex;
    box-shadow: rgba(0, 0, 0, 0.05) 0px 0px 0px 1px;
    align-items: center;
    justify-content: space-between;
    margin-bottom: 24px;
    border-radius: 15px;
    padding: 10px 20px;
    background-color: #fff;
}

.product-details-btn{
    text-decoration-line: none;
    color: #0031d1;
}

.product-details-btn:hover{
    text-decoration-line: underline;
    color: #022188;
}

.category-badge {
    display: inline-block;
    padding: 2px 10px;
    background-color: #007bff;
    color: white;
    border-radius: 20px;
    font-size: 0.85rem;
    font-weight: 500;
    text-transform: capitalize;
    letter-spacing: 0.5px;
    margin-left: 10px;
}

@media (max-width: 768px) {
    .category-badge {
        padding: 4px 10px;
        font-size: 0.75rem;
    }
}