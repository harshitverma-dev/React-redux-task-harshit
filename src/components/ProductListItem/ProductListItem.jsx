import React from 'react'
import { Link } from 'react-router-dom'
import './ProductListItem.css'
import { useDispatch } from 'react-redux'
import { setProductDetails } from '../../app/features/ListView/ProductSlice'

const ProductListItem = (props) => {
    const dispatch = useDispatch();
    const { productsItem } = props
    return (
        <div className="product-listing-item">
            <div className="product-listing-info">
                <h3 className="product-listing-title">{productsItem.title}</h3>
                <div className="product-listing-details">
                    <span className="price">${productsItem.price}</span>
                    <span className="category-badge">{productsItem.category.replace(/[-_]/g, ' ')}</span>
                </div>

            </div>
            <Link onClick={()=>dispatch(setProductDetails(productsItem))} to={`/product-detail/${productsItem.title.toLowerCase().replaceAll(' ', '-')}`} className="product-details-btn">
                Product Details
            </Link>
        </div>
    )
}

export default ProductListItem