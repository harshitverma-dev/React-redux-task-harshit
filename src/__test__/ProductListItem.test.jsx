import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import ProductListItem from '../components/ProductListItem/ProductListItem';
import productReducer from '../app/features/ListView/ProductSlice';

const createMockStore = (initialState = {}) => {
    return configureStore({
        reducer: {
            products: productReducer
        },
        preloadedState: initialState
    });
};

const renderWithProviders = (component, { store = createMockStore() } = {}) => {
    return render(
        <Provider store={store}>
            <BrowserRouter>
                {component}
            </BrowserRouter>
        </Provider>
    );
};

describe('ProductListItem Component', () => {
    const mockProduct = {
        id: 1,
        title: 'Test Product',
        price: 99.99,
        category: 'test_category',
        description: 'Test description'
    };

    it('renders product information correctly', () => {
        const { asFragment } = renderWithProviders(<ProductListItem productsItem={mockProduct} />);

        expect(screen.getByText('Test Product')).toBeInTheDocument();
        expect(screen.getByText('$99.99')).toBeInTheDocument();
        expect(screen.getByText('test category')).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('formats category text correctly', () => {
        const productWithComplexCategory = {
            ...mockProduct,
            category: 'test_category-name'
        };
        const { asFragment } = renderWithProviders(
            <ProductListItem productsItem={productWithComplexCategory} />
        );

        expect(screen.getByText('test category name')).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('creates correct product detail link', () => {
        const { asFragment } = renderWithProviders(<ProductListItem productsItem={mockProduct} />);
        
        const detailLink = screen.getByText('Product Details');
        expect(detailLink).toHaveAttribute(
            'href',
            '/product-detail/test-product'
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('dispatches setProductDetails action when clicking detail link', () => {
        const store = createMockStore();
        const dispatchSpy = vi.spyOn(store, 'dispatch');

        const { asFragment } = renderWithProviders(
            <ProductListItem productsItem={mockProduct} />, 
            { store }
        );

        fireEvent.click(screen.getByText('Product Details'));
        
        expect(dispatchSpy).toHaveBeenCalledWith(
            expect.objectContaining({
                type: 'productList/setProductDetails',
                payload: mockProduct
            })
        );
        expect(asFragment()).toMatchSnapshot();
    });
});