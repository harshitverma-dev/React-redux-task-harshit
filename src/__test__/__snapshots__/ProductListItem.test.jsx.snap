// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductListItem Component > creates correct product detail link 1`] = `
<DocumentFragment>
  <div
    class="product-listing-item"
  >
    <div
      class="product-listing-info"
    >
      <h3
        class="product-listing-title"
      >
        Test Product
      </h3>
      <div
        class="product-listing-details"
      >
        <span
          class="price"
        >
          $99.99
        </span>
        <span
          class="category-badge"
        >
          test category
        </span>
      </div>
    </div>
    <a
      class="product-details-btn"
      data-discover="true"
      href="/product-detail/test-product"
    >
      Product Details
    </a>
  </div>
</DocumentFragment>
`;

exports[`ProductListItem Component > dispatches setProductDetails action when clicking detail link 1`] = `
<DocumentFragment>
  <div
    class="product-listing-item"
  >
    <div
      class="product-listing-info"
    >
      <h3
        class="product-listing-title"
      >
        Test Product
      </h3>
      <div
        class="product-listing-details"
      >
        <span
          class="price"
        >
          $99.99
        </span>
        <span
          class="category-badge"
        >
          test category
        </span>
      </div>
    </div>
    <a
      class="product-details-btn"
      data-discover="true"
      href="/product-detail/test-product"
    >
      Product Details
    </a>
  </div>
</DocumentFragment>
`;

exports[`ProductListItem Component > formats category text correctly 1`] = `
<DocumentFragment>
  <div
    class="product-listing-item"
  >
    <div
      class="product-listing-info"
    >
      <h3
        class="product-listing-title"
      >
        Test Product
      </h3>
      <div
        class="product-listing-details"
      >
        <span
          class="price"
        >
          $99.99
        </span>
        <span
          class="category-badge"
        >
          test category name
        </span>
      </div>
    </div>
    <a
      class="product-details-btn"
      data-discover="true"
      href="/product-detail/test-product"
    >
      Product Details
    </a>
  </div>
</DocumentFragment>
`;

exports[`ProductListItem Component > renders product information correctly 1`] = `
<DocumentFragment>
  <div
    class="product-listing-item"
  >
    <div
      class="product-listing-info"
    >
      <h3
        class="product-listing-title"
      >
        Test Product
      </h3>
      <div
        class="product-listing-details"
      >
        <span
          class="price"
        >
          $99.99
        </span>
        <span
          class="category-badge"
        >
          test category
        </span>
      </div>
    </div>
    <a
      class="product-details-btn"
      data-discover="true"
      href="/product-detail/test-product"
    >
      Product Details
    </a>
  </div>
</DocumentFragment>
`;
